# ETL Notebook 
This notebook documents my step-by-step thinking and justification behind my decisions for building the ETL pipeline.

## EXTRACT

In [60]:
from pathlib import Path
import pandas as pd
import ast

Each athlete lives in a 'p[XX]/pmsys/' folder. I glob 'p[0-9][0-9]' so the loop picks up only athlete folders and never the 'participant-overview.xlsx', and 'sorted()' makes athlete order deterministic. This way p01 always processes first, so generated IDs are reproducible across runs.

Two absence cases are handled differently on purpose. p08 has no 'injury.csv' file at all, so the injury read is guarded and p08 simply contributes zero injury rows (a missing file is valid, not an error). 'srpe.csv' and 'wellness.csv' are present for every athlete, so those are intentionally unguarded. A missing srpe or wellness file would be a genuine surprise and should halt the run loudly rather than be silently swallowed.

In [61]:
PMDATA_DIR = Path("data/pmdata")

injury_frames, srpe_frames, wellness_frames = [], [], []

# p[0-9][0-9] matches p01..p16 only — NOT participant-overview.xlsx
# sorted() keeps them in order
for athlete_dir in sorted(PMDATA_DIR.glob("p[0-9][0-9]")):
    source_id = athlete_dir.name # e.g. "p01"
    pmsys = athlete_dir / "pmsys"

    # srpe + wellness are present for every athlete
    srpe = pd.read_csv(pmsys / "srpe.csv", parse_dates=["end_date_time"])
    srpe.insert(0, "source_id", source_id)
    srpe_frames.append(srpe)

    wellness = pd.read_csv(pmsys / "wellness.csv", parse_dates=["effective_time_frame"])
    wellness.insert(0, "source_id", source_id)
    wellness_frames.append(wellness)

    # guard: p08 has no injury.csv
    injury_path = pmsys / "injury.csv"
    if injury_path.exists():
        injury = pd.read_csv(injury_path, parse_dates=["effective_time_frame"])
        injury.insert(0, "source_id", source_id)
        injury_frames.append(injury)
    else:
        print(f"skipping injury for {source_id}: {injury_path} not found")

injury = pd.concat(injury_frames, ignore_index=True)
srpe = pd.concat(srpe_frames, ignore_index=True)
wellness = pd.concat(wellness_frames, ignore_index=True)

for name, df in {"injury": injury, "srpe": srpe, "wellness": wellness}.items():
    print(f"{name}: {df.shape[0]} rows x {df.shape[1]} cols, "
          f"{df['source_id'].nunique()} athletes")

skipping injury for p08: data\pmdata\p08\pmsys\injury.csv not found
injury: 225 rows x 3 cols, 15 athletes
srpe: 783 rows x 5 cols, 16 athletes
wellness: 1747 rows x 10 cols, 16 athletes


Sanity check on participant_overview df

In [62]:
participant_overview = pd.read_excel(PMDATA_DIR / "participant-overview.xlsx", header=1)
participant_overview

,Participant ID,Age,Height,Gender,A or B person,Max heart rate,Date,Minutes,Seconds,Stride walk,Stride run
0,p01,48,195,male,A,182,2019-11-26 00:00:00,29,33,80.90,102.9
1,p02,60,180,male,A,169,2019-12-15 00:00:00,23,51,74.70,92.4
2,p03,25,184,male,A,157,2019-12-30 00:00:00,33,22,NaN,NaN
3,p04,26,163,female,A,195,2019-11-19 00:00:00,22,13,67.30,110.2
4,p05,35,176,male,A,184,2019-12-23 00:00:00,32,40,73.00,94.3
5,p06,42,179,male,B,181,2019-12-01 00:00:00,23,19,73.04,97.6
6,p07,26,177,male,B,NaN,2019-11-19 00:00:00,19,40,73.50,119.5
7,p08,27,186,male,B,200,2019-11-28 00:00:00,18,47,77.20,103.6
8,p09,26,180,male,B,183,2020-01-07 00:00:00,35,6,74.70,109.9
9,p10,38,179,female,B,197,2019-12-08 00:00:00,28,10,73.09,102.3


Decided to only retain demographic information about each athlete:

In [63]:
athlete = participant_overview.drop(columns=['Date', 'Minutes', 'Seconds', 'Stride walk', 'Stride run'])
athlete

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182
1,p02,60,180,male,A,169
2,p03,25,184,male,A,157
3,p04,26,163,female,A,195
4,p05,35,176,male,A,184
5,p06,42,179,male,B,181
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200
8,p09,26,180,male,B,183
9,p10,38,179,female,B,197


I noticed p08 had no injury.csv file and wanted to inspect how many injuries were being reported and how many of them were empty per athlete:

In [64]:
injury["is_empty"] = injury["injuries"].isin([{}, "{}"])   # dict vs string, see below
print(injury.groupby("source_id")["is_empty"].agg(["sum", "count"]))

           sum  count
source_id            
p01         23     24
p02          4      5
p03         11     11
p04         14     17
p05          0     10
p06         15     15
p07         13     15
p09         20     20
p10         12     12
p11          1      4
p12          1     49
p13         16     16
p14          7     10
p15          0      6
p16         11     11


This tells us that there aren't clean periodic check-ins. Otherwise we would have "{}" - empty injury check-ins on days where the athlete is healthy and injury free.  

Next, I want to see if there are persisting injuries being logged repeatedly. Let's take p12 for example and see what injuries are being reported:

In [65]:
p12 = injury[injury["source_id"] == "p12"]
p12[p12["injuries"] != "{}"][["effective_time_frame", "injuries"]].head(20)

,effective_time_frame,injuries
133,2019-11-15 20:01:20.647000+00:00,{'right_knee': 'minor'}
135,2019-11-18 04:50:00.666000+00:00,{'right_knee': 'minor'}
136,2019-11-18 06:01:40.574000+00:00,{'right_knee': 'minor'}
137,2019-11-18 19:26:24.936000+00:00,{'left_shoulder': 'minor'}
138,2019-11-19 05:49:39.091000+00:00,"{'left_leg': 'minor', 'left_shoulder': 'minor'..."
139,2019-11-20 05:58:09.329000+00:00,"{'left_leg': 'minor', 'left_shoulder': 'minor'..."
140,2019-11-20 20:03:10.579000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
141,2019-11-21 20:59:32.959000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
142,2019-11-22 22:35:26.660000+00:00,"{'left_leg': 'minor', 'right_knee': 'minor'}"
143,2019-11-24 10:10:14.803000+00:00,{'right_knee': 'minor'}


Note that there are multiple injuries being reported on the same day. These will need to be exploded so that a single row turns into N-number (number of injured areas per day) of rows so there is one per body part, each with that part's severity.

## TRANSFORM

### Injury: grain and the report-vs-episode decision

A single row in the 'injury.csv' file is a one weekly check-in, not one injury. '{}' means that "checked in, no injury" which is useful information indicating the athlete was healthy that week. Hence, these rows don't belong in an injury fact table. Non-empty rows hold a '{injury_area: severity}' dict, and one row can name several body parts, so I explode to one row per (athlete, timestamp, injury_area, severity).

Key decision: I chose to keep the full timestamp rather than collapsing to a date. Several athletes re-report the *same* injury across consecutive check-ins (p12's right knee spans weeks). Collapsing to a date would manufacture false duplicates and, worse, counting rows would wildly overstate injury counts. So the fact table stays at report grain which is lossless and faithful to the source. The "collapse repeated reports into injury episodes" is left as a downstream derivation, not baked into the load.

'injuries' arrives from 'read_csv' as a string and not a dict. I parse with 'ast.literal_eval' and confirm every row parses to a dict before exploding.

In [66]:
injury["injury_dict"] = injury["injuries"].apply(ast.literal_eval) # Convert injuries column to dicts

In [67]:
injured = injury.loc[injury["injury_dict"] != {}, ["source_id", "effective_time_frame", "injury_dict"]]
injured

,source_id,effective_time_frame,injury_dict
11,p01,2020-01-07 22:33:38.989000+00:00,{'right_hand': 'minor'}
24,p02,2019-11-13 13:53:01.675000+00:00,"{'left_elbow': 'minor', 'right_shoulder': 'min..."
43,p04,2019-11-26 06:24:57.956000+00:00,{'groin_hip': 'minor'}
50,p04,2020-01-08 08:03:03.131000+00:00,"{'left_leg': 'minor', 'right_leg': 'minor'}"
53,p04,2020-01-26 21:34:24.956000+00:00,"{'left_leg': 'minor', 'right_leg': 'minor'}"
...,...,...,...
209,p15,2020-02-10 08:05:01.763000+00:00,{'stomach_back': 'major'}
210,p15,2020-02-11 08:10:36.015000+00:00,{'stomach_back': 'minor'}
211,p15,2020-02-13 05:46:38.880000+00:00,{'stomach_back': 'major'}
212,p15,2020-02-15 07:22:56.107000+00:00,{'stomach_back': 'major'}


In [68]:
records = []
for row in injured.itertuples():
    ath_id = row.source_id
    effective_time_frame = row.effective_time_frame
    inj = row.injury_dict
    for injury_area, severity in inj.items():
        records.append({
            "source_id": ath_id,
            "effective_time_frame": effective_time_frame,
            "injury_area": injury_area,
            "severity": severity,
        })
injuries = pd.DataFrame(records)

In [69]:
injuries.shape

(111, 4)

In [70]:
injuries["injury_area"].value_counts()

injury_area
right_knee        35
left_foot         15
left_leg          11
right_foot         9
left_knee          5
left_hand          5
left_shoulder      5
stomach_back       5
chest              4
right_hand         4
head_neck          3
right_leg          3
groin_hip          2
right_shoulder     2
left_elbow         1
right_thigh        1
right_elbow        1
Name: count, dtype: int64

In [71]:
injuries["effective_time_frame"] = injuries.effective_time_frame.dt.floor("s").dt.tz_localize(None)
injuries

,source_id,effective_time_frame,injury_area,severity
0,p01,2020-01-07 22:33:38,right_hand,minor
1,p02,2019-11-13 13:53:01,left_elbow,minor
2,p02,2019-11-13 13:53:01,right_shoulder,minor
3,p04,2019-11-26 06:24:57,groin_hip,minor
4,p04,2020-01-08 08:03:03,left_leg,minor
...,...,...,...,...
106,p15,2020-02-10 08:05:01,stomach_back,major
107,p15,2020-02-11 08:10:36,stomach_back,minor
108,p15,2020-02-13 05:46:38,stomach_back,major
109,p15,2020-02-15 07:22:56,stomach_back,major


Once exploded, (source_id, timestamp, injury_area) is unique - verified with a duplicate check returning 0. I still add a surrogate 'injury_id' as the primary key: a single-column integer key gives clean joins, and the natural composite key becomes a UNIQUE constraint that documents the grain guarantee.

In [72]:
duplicate_rows = injuries[injuries.duplicated(subset=["source_id", "effective_time_frame", "injury_area"])]
print(len(duplicate_rows))

0


Adding a surrogate key for injury (injury_id):

In [73]:
injuries = injuries.reset_index(drop=True)
injuries.insert(0, "injury_id", range(1, len(injuries) + 1))
injuries

,injury_id,source_id,effective_time_frame,injury_area,severity
0,1,p01,2020-01-07 22:33:38,right_hand,minor
1,2,p02,2019-11-13 13:53:01,left_elbow,minor
2,3,p02,2019-11-13 13:53:01,right_shoulder,minor
3,4,p04,2019-11-26 06:24:57,groin_hip,minor
4,5,p04,2020-01-08 08:03:03,left_leg,minor
...,...,...,...,...,...
106,107,p15,2020-02-10 08:05:01,stomach_back,major
107,108,p15,2020-02-11 08:10:36,stomach_back,minor
108,109,p15,2020-02-13 05:46:38,stomach_back,major
109,110,p15,2020-02-15 07:22:56,stomach_back,major


In [74]:
print(injuries.dtypes)

injury_id                        int64
source_id                       object
effective_time_frame    datetime64[ns]
injury_area                     object
severity                        object
dtype: object


### Wellness: dropping empty submissions

The grain for the wellness csv is different to the injury one. Some athletes submit wellness more than once a day, and inspection shows these are genuinely different reports (morning vs evening values differ), not duplications. So (athlete, day) is not an appropriate grain and would cause the same mistake as collapsing injury to a date. I store wellness at report grain, keyed on UNIQUE KEY (athlete_id, reported_at); the load succeeding proves uniqueness at the second precision. I defer the one-row-per-day collapse to a SQL view (first report of day). Cleaning stays reversible and the base table stays faithful. 

I want to examine if there are multiple wellness reports on the same day:

In [75]:
multi = wellness.groupby(["source_id",wellness["effective_time_frame"].dt.date]).filter(lambda g: len(g) > 1)
multi.sort_values(["source_id", "effective_time_frame"])

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress
257,p03,2019-12-01 07:38:02.291000+00:00,3,3,5,6,3,3,[],3
258,p03,2019-12-01 21:12:20.093000+00:00,3,3,5,5,3,3,[],3
261,p03,2019-12-11 06:41:11.728000+00:00,3,3,5,7,3,3,[],3
262,p03,2019-12-11 17:01:26.334000+00:00,3,3,5,7,3,3,[],3
278,p03,2020-01-12 00:32:08.115000+00:00,3,3,5,7,3,3,[],3
279,p03,2020-01-12 15:28:39.900000+00:00,3,3,5,7,3,3,[],3
297,p03,2020-02-02 01:00:28.003000+00:00,3,3,5,8,3,3,[],3
298,p03,2020-02-02 18:15:39.697000+00:00,3,3,5,8,3,3,[],3
439,p04,2020-03-07 08:17:42.897000+00:00,3,3,5,8,4,3,[],3
440,p04,2020-03-07 08:18:13.008000+00:00,3,3,5,8,3,3,[],3


This is an interesting find since the paper states that fatigue, mood, sleep_quality, soreness and stress are rated on a 1-5 scale while readiness is rated 0-10. A row where all five are 0 is therefore an empty/aborted submission, not a real report. I drop only those fully-empty rows (`.all(axis=1)`, not `.any()`): rows missing a single field are genuine reports with one blank value and are kept. Verified this removes exactly the expected rows.

In [76]:
five_pt_cols = ['fatigue', 'mood', 'sleep_quality', 'soreness', 'stress']
wellness[wellness[five_pt_cols].eq(0).all(axis=1)]

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress
502,p05,2019-12-12 10:50:32.079000+00:00,0,0,0,0,0,0,[],0
510,p05,2019-12-20 06:58:36.783000+00:00,0,0,0,0,0,0,[],0


In [77]:
empty_mask = wellness[five_pt_cols].eq(0).all(axis=1)
wellness = wellness[~empty_mask]
wellness = wellness.reset_index(drop=True)

Now I want to check if any of these columns have a 0 to see how to handle this case:

In [78]:
wellness[wellness[five_pt_cols].eq(0).any(axis=1)]

,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress
373,p04,2019-12-27 09:00:21.074000+00:00,2,3,3,6,2,1,"[48979004, 32696007, 12921003, 22943007, 36820...",0
1201,p11,2019-12-06 08:39:06.697000+00:00,2,0,5,6,3,3,[],2
1406,p13,2019-12-25 17:05:13.971000+00:00,2,3,3,6,2,2,[12921003],0


Some fields are full but others are at 0. This tells me that the rows are still real reports and the athlete has just not logged 'stress' or 'mood'. So I chose to drop rows where all five are 0 but treated a lone zero as a missing value and a reason to not drop the entire row.

I'm also adding wellness_id here so every value in soreneess_report can be proved to exist in wellness so that there are no orphans possible or parallel numbering to drift.

In [79]:
wellness.insert(0, "wellness_id", range(1, len(wellness) + 1))
wellness["effective_time_frame"] = wellness.effective_time_frame.dt.floor("s").dt.tz_localize(None)
wellness

,wellness_id,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress
0,1,p01,2019-11-01 08:31:40,2,3,5,6,3,2,[12921003],3
1,2,p01,2019-11-02 10:00:01,2,3,6,6,3,2,[12921003],3
2,3,p01,2019-11-03 14:28:03,3,3,8,6,3,3,[],3
3,4,p01,2019-11-04 07:05:28,3,3,8,6,3,3,[],3
4,5,p01,2019-11-05 06:13:35,3,3,8,5,3,3,[],3
...,...,...,...,...,...,...,...,...,...,...,...
1740,1741,p16,2020-03-08 11:44:44,3,3,4,6,2,3,[],3
1741,1742,p16,2020-03-11 11:33:15,2,3,4,6,2,3,[],2
1742,1743,p16,2020-03-15 08:33:15,4,4,4,10,4,4,[],4
1743,1744,p16,2020-03-17 09:54:27,2,3,4,5,3,3,[],3


### Training sessions: dropping empty submissions and splitting the activity list

'activity_names' blends two different concepts in one list e.g. '['individual', 'strength']'. It consists of participation *mode* with an activity *type*. I split them: exactly one mode per session (verified across all sessions) becomes a scalar 'participation_mode' column on 'training_session', while the 1-to-many activity types explode into a separate 'session_activity' bridge table. 'session_load' is materialized here as RPE x duration - a derived column I store deliberately because it's the workload primitive ACWR sums over, and I'd rather compute it once than recompute it in every window query.
 
 Next, I want to see if the participation mode ("individual" or "team") is being reported once per activity in the srpe.csv file: 

In [80]:
srpe["activity_list"] = srpe["activity_names"].apply(ast.literal_eval)
srpe["activity_list"].apply(lambda l: len([x for x in l if x in {"individual","team"}])).value_counts()

activity_list
1    772
0     11
Name: count, dtype: int64

772 rows have the participation mode in the activity name and 11 don't. I also want to see the data type distribution and check for the entries that don't have participation mode:

In [81]:
srpe["activity_list"].apply(type).value_counts()

activity_list
<class 'list'>    772
<class 'dict'>     11
Name: count, dtype: int64

In [82]:
srpe[srpe["activity_list"].apply(type) == dict][["end_date_time", "activity_names", "activity_list"]]

,end_date_time,activity_names,activity_list
772,2019-11-18 13:34:02.327000+00:00,{},{}
773,2019-11-25 14:25:04.577000+00:00,{},{}
774,2019-12-09 12:32:35.414000+00:00,{},{}
775,2019-12-23 16:12:08.718000+00:00,{},{}
776,2019-12-30 10:29:31.987000+00:00,{},{}
777,2020-01-13 15:37:07.243000+00:00,{},{}
778,2020-01-20 15:42:00.634000+00:00,{},{}
779,2020-01-28 20:53:03.947000+00:00,{},{}
780,2020-02-04 21:43:29.297000+00:00,{},{}
781,2020-02-10 15:29:39.385000+00:00,{},{}


In [83]:
srpe[srpe["activity_list"].apply(type) == dict][["perceived_exertion", "duration_min", "activity_list"]]

,perceived_exertion,duration_min,activity_list
772,NaN,NaN,{}
773,NaN,NaN,{}
774,NaN,NaN,{}
775,NaN,NaN,{}
776,NaN,NaN,{}
777,NaN,NaN,{}
778,NaN,NaN,{}
779,NaN,NaN,{}
780,NaN,NaN,{}
781,NaN,NaN,{}


11 srpe rows have an empty `{}` activity *and* NaN RPE and NaN duration - no activity, no load, nothing computable. Unlike injury `{}` (which carried information), these carry none, so I drop them. I filter on "keep rows that parsed to a list" rather than "drop the dict rows. Keeping the known-good shape is safer than blacklisting the one bad shape I happened to find.

In [84]:
before = len(srpe)
srpe = srpe[srpe["activity_list"].apply(type) == list]   # keep only the list rows
srpe = srpe.reset_index(drop=True)
print(before - len(srpe))   # expect 11

11


I am adding session_id to the srpe table so it carries over to the child tables

In [85]:
srpe.insert(0, "session_id", range(1, len(srpe) + 1))
srpe

,session_id,source_id,end_date_time,activity_names,perceived_exertion,duration_min,activity_list
0,1,p01,2019-11-05 22:51:54.710000+00:00,"['individual', 'running']",7.0,30.0,"[individual, running]"
1,2,p01,2019-11-11 21:15:15.092000+00:00,"['individual', 'running']",6.0,30.0,"[individual, running]"
2,3,p01,2019-11-14 21:00:53+00:00,"['team', 'soccer']",7.0,60.0,"[team, soccer]"
3,4,p01,2019-11-18 21:34:44.995000+00:00,"['individual', 'endurance']",6.0,50.0,"[individual, endurance]"
4,5,p01,2019-11-21 21:06:25.146000+00:00,"['team', 'soccer']",6.0,70.0,"[team, soccer]"
...,...,...,...,...,...,...,...
767,768,p15,2020-03-24 12:30:25+00:00,"['individual', 'endurance']",6.0,100.0,"[individual, endurance]"
768,769,p15,2020-03-25 15:30:39+00:00,"['individual', 'endurance']",4.0,90.0,"[individual, endurance]"
769,770,p15,2020-03-26 13:00:39+00:00,"['individual', 'endurance']",6.0,120.0,"[individual, endurance]"
770,771,p15,2020-03-27 15:59:02+00:00,"['individual', 'endurance']",3.0,100.0,"[individual, endurance]"


I want to change the dtypes for perceived_exertion and duration_min to int64 to match the DDL.

In [86]:
srpe['perceived_exertion'] = srpe['perceived_exertion'].astype(int)
srpe['duration_min'] = srpe['duration_min'].astype(int)

In [87]:
training_records = []
for row in srpe.itertuples():
    activity_list = row.activity_list
    mode = [x for x in activity_list if x in {"individual", "team"}][0]

    training_records.append({
        "session_id": row.session_id,
        "source_id": row.source_id, 
        "end_time": row.end_date_time,
        "rpe": row.perceived_exertion,
        "duration_min": row.duration_min,
        "session_load": row.perceived_exertion * row.duration_min,
        "participation_mode": mode
    })

training_session = pd.DataFrame(training_records)
training_session['end_time'] = training_session['end_time'].dt.floor("s").dt.tz_localize(None)
training_session

,session_id,source_id,end_time,rpe,duration_min,session_load,participation_mode
0,1,p01,2019-11-05 22:51:54,7,30,210,individual
1,2,p01,2019-11-11 21:15:15,6,30,180,individual
2,3,p01,2019-11-14 21:00:53,7,60,420,team
3,4,p01,2019-11-18 21:34:44,6,50,300,individual
4,5,p01,2019-11-21 21:06:25,6,70,420,team
...,...,...,...,...,...,...,...
767,768,p15,2020-03-24 12:30:25,6,100,600,individual
768,769,p15,2020-03-25 15:30:39,4,90,360,individual
769,770,p15,2020-03-26 13:00:39,6,120,720,individual
770,771,p15,2020-03-27 15:59:02,3,100,300,individual


In [88]:
print(training_session.dtypes)

session_id                     int64
source_id                     object
end_time              datetime64[ns]
rpe                            int64
duration_min                   int64
session_load                   int64
participation_mode            object
dtype: object


### session_activity:

'session_activity' and 'training_session' both inherit the 'session_id' from the 'srpe' table I built earlier. This ensures that original order is preserved and the child tables inherit the correct ids from the parent table.

In [89]:
activity_records = []
for row in srpe.itertuples():
    activity_list = row.activity_list
    activities = [x for x in activity_list if x not in {"individual", "team"}]
    for activity in activities:
        activity_records.append({
            "session_id": row.session_id,
            "activity_type": activity,
        })
session_activity = pd.DataFrame(activity_records)
session_activity

,session_id,activity_type
0,1,running
1,2,running
2,3,soccer
3,4,endurance
4,5,soccer
...,...,...
836,768,endurance
837,769,endurance
838,770,endurance
839,771,endurance


In [90]:
session_activity["activity_type"].value_counts()

activity_type
running      402
endurance    257
strength     170
soccer        12
Name: count, dtype: int64

### Working on the soreness_report and body_part lookup tables:

soreness_area is a repeating group. A single report in wellness can list multiple body parts which violates first normal form. This causes it to explode to a child table at one row per (submission, body part). The parent keeps its scalar soreness severity score, the child holds the body parts.

In [91]:
wellness

,wellness_id,source_id,effective_time_frame,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,soreness_area,stress
0,1,p01,2019-11-01 08:31:40,2,3,5,6,3,2,[12921003],3
1,2,p01,2019-11-02 10:00:01,2,3,6,6,3,2,[12921003],3
2,3,p01,2019-11-03 14:28:03,3,3,8,6,3,3,[],3
3,4,p01,2019-11-04 07:05:28,3,3,8,6,3,3,[],3
4,5,p01,2019-11-05 06:13:35,3,3,8,5,3,3,[],3
...,...,...,...,...,...,...,...,...,...,...,...
1740,1741,p16,2020-03-08 11:44:44,3,3,4,6,2,3,[],3
1741,1742,p16,2020-03-11 11:33:15,2,3,4,6,2,3,[],2
1742,1743,p16,2020-03-15 08:33:15,4,4,4,10,4,4,[],4
1743,1744,p16,2020-03-17 09:54:27,2,3,4,5,3,3,[],3


soreness_area arrives as a string and I parse it into a new soreness_codes column rather than overwriting. This allows the cell to be re-runnable and the raw payload survives as audit trail. I check the parsed type across all rows. len() succeeds on both lists and dicts, so a dict-shaped row would sit  in the zero bucket of the length distribution undetected. This is the same check that found the 11 malformed rows in srpe.

In [ ]:
wellness["soreness_codes"] = wellness["soreness_area"].apply(ast.literal_eval)

list

In [93]:
print(wellness.soreness_codes.apply(len).value_counts())
print(wellness["soreness_codes"].apply(type).value_counts())

soreness_codes
0    1370
2     133
1      98
3      88
4      33
5      12
6       6
7       5
Name: count, dtype: int64
soreness_codes
<class 'list'>    1745
Name: count, dtype: int64


The exploded soreness_code frequencies establish that only 7 distinct codes are used in this dataset. These codes need to go into the lookup table.

In [94]:
wellness.soreness_codes.explode().dropna().value_counts()

soreness_codes
32696007     228
48979004     224
12921003     185
69536005      85
368208006     61
368209003     59
22943007      49
Name: count, dtype: int64

soreness_report uses a composite primary key of (wellness_id, body_part_code), which is only possible if no submission names the same code twice. A repeat would make the natural key non-unique and force a surrogate. A zero on this next check permits the composite PK, so the constraint is a claim the data supports rather than an assumption.

In [95]:
wellness.soreness_codes.apply(lambda l: len(l) != len(set(l))).sum()

0

The explode reads wellness_id directly off the parent row it came from to uphold referential integrity by construction. Every child key is a value that provably exists in the parent. The alternative is generating a second numbering and trusting two independent sequences to agree. session_activity originally did that, and I refactored it to this pattern.

In [96]:
soreness_records = []
for row in wellness.itertuples():
    sore_part = row.soreness_codes
    for area in sore_part:
        soreness_records.append({
            "wellness_id": row.wellness_id,
            "body_part_code": area,
        })
soreness_report = pd.DataFrame(soreness_records)
soreness_report

,wellness_id,body_part_code
0,1,12921003
1,2,12921003
2,6,48979004
3,6,32696007
4,11,48979004
...,...,...
886,1663,22943007
887,1663,69536005
888,1668,69536005
889,1725,368208006


The exploded codes join to a body_part lookup. SNOMED-CT codes are assigned by an external standards body, globally unique and never reissued which allows the code to become the natural primary key. The names are hand-mapped from the SNOMED-CT browser to their integer codes (int64). This is the one mapping in the project that nothing downstream can validate; a left / right transposition would be silent and undetectable. 

In [97]:
body_part_lookup = {
    'body_part_code': [12921003, 48979004, 32696007, 22943007, 368208006, 368209003, 69536005],
    'body_part_name': ['Pelvis', 'Left lower leg', 'Right lower leg', 'Torso', 'Left upper arm', 'Right upper arm', 'Head']
}

body_part = pd.DataFrame(body_part_lookup)
body_part


,body_part_code,body_part_name
0,12921003,Pelvis
1,48979004,Left lower leg
2,32696007,Right lower leg
3,22943007,Torso
4,368208006,Left upper arm
5,368209003,Right upper arm
6,69536005,Head


### Athlete dimension and the ID bridge

The athlete dimension is the hub every fact references, so it's built first. It comes from 'participant-overview.xlsx', which has all 16 athletes including p08 who has demographics, wellness and srpe but no injuries. The dimension is the complete roster and the facts are sparse against it.

max_heart_rate arrives with missing values as both blanks and NaNs; I coerce the column with 'pd.to_numeric(errors="coerce")' so all missing forms become NaN (NULL in SQL). I don't fill them since NULL is an honest representation of "not measured". I keep both 'athlete_id' (integer surrogate PK) and 'source_id' ("p01" is a UNIQUE natural key) to create a bridge every fact maps through.

In [98]:
athlete

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182
1,p02,60,180,male,A,169
2,p03,25,184,male,A,157
3,p04,26,163,female,A,195
4,p05,35,176,male,A,184
5,p06,42,179,male,B,181
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200
8,p09,26,180,male,B,183
9,p10,38,179,female,B,197


In [99]:
athlete["Max heart rate"] = pd.to_numeric(athlete["Max heart rate"], errors = "coerce")
athlete["Gender"] = athlete["Gender"].str.strip()
athlete

,Participant ID,Age,Height,Gender,A or B person,Max heart rate
0,p01,48,195,male,A,182.0
1,p02,60,180,male,A,169.0
2,p03,25,184,male,A,157.0
3,p04,26,163,female,A,195.0
4,p05,35,176,male,A,184.0
5,p06,42,179,male,B,181.0
6,p07,26,177,male,B,NaN
7,p08,27,186,male,B,200.0
8,p09,26,180,male,B,183.0
9,p10,38,179,female,B,197.0


In [100]:
athlete = athlete.rename(columns={
    'Participant ID': 'source_id',
    'Age': 'age',
    'Height': 'height_cm',
    'Gender': 'gender',
    'A or B person': 'personality_type',
    'Max heart rate': 'max_heart_rate'
    })
athlete["max_heart_rate"] = athlete["max_heart_rate"].astype("Int64")
athlete = athlete.reset_index(drop=True)
athlete.insert(0, "athlete_id", range(1, len(athlete) + 1))
athlete


,athlete_id,source_id,age,height_cm,gender,personality_type,max_heart_rate
0,1,p01,48,195,male,A,182
1,2,p02,60,180,male,A,169
2,3,p03,25,184,male,A,157
3,4,p04,26,163,female,A,195
4,5,p05,35,176,male,A,184
5,6,p06,42,179,male,B,181
6,7,p07,26,177,male,B,<NA>
7,8,p08,27,186,male,B,200
8,9,p09,26,180,male,B,183
9,10,p10,38,179,female,B,197


Checking data type and verifying two-value enums for proper handling:

In [101]:
print(athlete["personality_type"].value_counts())
print(athlete["gender"].value_counts())

personality_type
A    10
B     6
Name: count, dtype: int64
gender
male      13
female     3
Name: count, dtype: int64


Every df carried 'source_id' through transformation. Now that the athlete dimension defines source_id -> athlete_id, I map each df's 'source_id' to its integer 'athlete_id' to satisfy the foreign keys, and check that no mapped key is null (every df must reference an athlete that exists). 'source_id' is dropped from the dfs before load and it lives only on the athlete table.

In [102]:
id_map = athlete.set_index("source_id")["athlete_id"]
injuries["athlete_id"] = injuries["source_id"].map(id_map)
wellness["athlete_id"] = wellness["source_id"].map(id_map)
training_session["athlete_id"] = training_session["source_id"].map(id_map)

In [103]:
injuries = injuries.drop(columns=['source_id'])
wellness = wellness.drop(columns=['source_id', 'soreness_area', 'soreness_codes'])
training_session = training_session.drop(columns=['source_id'])

In [104]:
print(injuries["athlete_id"].isna().sum())# expect 0
print(wellness["athlete_id"].isna().sum()) # expect 0
print(training_session["athlete_id"].isna().sum()) # expect 0

0
0
0


In [106]:
injuries = injuries.rename(columns={"effective_time_frame": "reported_at"})
injuries

,injury_id,reported_at,injury_area,severity,athlete_id
0,1,2020-01-07 22:33:38,right_hand,minor,1
1,2,2019-11-13 13:53:01,left_elbow,minor,2
2,3,2019-11-13 13:53:01,right_shoulder,minor,2
3,4,2019-11-26 06:24:57,groin_hip,minor,4
4,5,2020-01-08 08:03:03,left_leg,minor,4
...,...,...,...,...,...
106,107,2020-02-10 08:05:01,stomach_back,major,15
107,108,2020-02-11 08:10:36,stomach_back,minor,15
108,109,2020-02-13 05:46:38,stomach_back,major,15
109,110,2020-02-15 07:22:56,stomach_back,major,15


Verifying that severity is only being reported as 'minor' and 'major':

In [107]:
print(injuries["severity"].value_counts())

severity
minor    107
major      4
Name: count, dtype: int64


In [109]:
wellness = wellness.rename(columns={"effective_time_frame": "reported_at"})
wellness

,wellness_id,reported_at,fatigue,mood,readiness,sleep_duration_h,sleep_quality,soreness,stress,athlete_id
0,1,2019-11-01 08:31:40,2,3,5,6,3,2,3,1
1,2,2019-11-02 10:00:01,2,3,6,6,3,2,3,1
2,3,2019-11-03 14:28:03,3,3,8,6,3,3,3,1
3,4,2019-11-04 07:05:28,3,3,8,6,3,3,3,1
4,5,2019-11-05 06:13:35,3,3,8,5,3,3,3,1
...,...,...,...,...,...,...,...,...,...,...
1740,1741,2020-03-08 11:44:44,3,3,4,6,2,3,3,16
1741,1742,2020-03-11 11:33:15,2,3,4,6,2,3,2,16
1742,1743,2020-03-15 08:33:15,4,4,4,10,4,4,4,16
1743,1744,2020-03-17 09:54:27,2,3,4,5,3,3,3,16


Pre-load contract check: seven frames, in the order they load. Parents precede children so every foreign key has a target by the time it's inserted.

In [110]:
frames = {
    "athlete":          athlete,
    "body_part":        body_part,
    "injuries":         injuries,
    "wellness":         wellness,
    "training_session": training_session,
    "soreness_report":  soreness_report,
    "session_activity": session_activity,
}

for name, df in frames.items():
    print(f"{name:18} {str(df.shape):12} {df.columns.tolist()}")

athlete            (16, 7)      ['athlete_id', 'source_id', 'age', 'height_cm', 'gender', 'personality_type', 'max_heart_rate']
body_part          (7, 2)       ['body_part_code', 'body_part_name']
injuries           (111, 5)     ['injury_id', 'reported_at', 'injury_area', 'severity', 'athlete_id']
wellness           (1745, 10)   ['wellness_id', 'reported_at', 'fatigue', 'mood', 'readiness', 'sleep_duration_h', 'sleep_quality', 'soreness', 'stress', 'athlete_id']
training_session   (772, 7)     ['session_id', 'end_time', 'rpe', 'duration_min', 'session_load', 'participation_mode', 'athlete_id']
soreness_report    (891, 2)     ['wellness_id', 'body_part_code']
session_activity   (841, 2)     ['session_id', 'activity_type']


---

# LOAD

In [111]:
import os
from dotenv import load_dotenv
load_dotenv()
from sqlalchemy import create_engine, text

In [112]:
pw = os.environ.get("MYSQL_PASSWORD")
engine = create_engine(f"mysql+pymysql://root:{pw}@localhost:3306/acwr")
with engine.connect() as connection:
    result = connection.execute(text("SELECT VERSION()"))
    print(result.fetchone())

('8.0.43',)


In [113]:
print(athlete.to_sql("athlete", engine, if_exists="append", index=False))
print(body_part.to_sql("body_part", engine, if_exists="append", index=False))
print(injuries.to_sql("injuries", engine, if_exists="append", index=False))
print(wellness.to_sql("wellness", engine, if_exists="append", index=False))
print(training_session.to_sql("training_session", engine, if_exists="append", index=False))
print(soreness_report.to_sql("soreness_report", engine, if_exists="append", index=False))
print(session_activity.to_sql("session_activity", engine, if_exists="append", index=False))

IntegrityError: (pymysql.err.IntegrityError) (1062, "Duplicate entry '1' for key 'athlete.PRIMARY'")
[SQL: INSERT INTO athlete (athlete_id, source_id, age, height_cm, gender, personality_type, max_heart_rate) VALUES (%(athlete_id)s, %(source_id)s, %(age)s, %(height_cm)s, %(gender)s, %(personality_type)s, %(max_heart_rate)s)]
[parameters: [{'athlete_id': 1, 'source_id': 'p01', 'age': 48, 'height_cm': 195, 'gender': 'male', 'personality_type': 'A', 'max_heart_rate': 182}, {'athlete_id': 2, 'source_id': 'p02', 'age': 60, 'height_cm': 180, 'gender': 'male', 'personality_type': 'A', 'max_heart_rate': 169}, {'athlete_id': 3, 'source_id': 'p03', 'age': 25, 'height_cm': 184, 'gender': 'male', 'personality_type': 'A', 'max_heart_rate': 157}, {'athlete_id': 4, 'source_id': 'p04', 'age': 26, 'height_cm': 163, 'gender': 'female', 'personality_type': 'A', 'max_heart_rate': 195}, {'athlete_id': 5, 'source_id': 'p05', 'age': 35, 'height_cm': 176, 'gender': 'male', 'personality_type': 'A', 'max_heart_rate': 184}, {'athlete_id': 6, 'source_id': 'p06', 'age': 42, 'height_cm': 179, 'gender': 'male', 'personality_type': 'B', 'max_heart_rate': 181}, {'athlete_id': 7, 'source_id': 'p07', 'age': 26, 'height_cm': 177, 'gender': 'male', 'personality_type': 'B', 'max_heart_rate': None}, {'athlete_id': 8, 'source_id': 'p08', 'age': 27, 'height_cm': 186, 'gender': 'male', 'personality_type': 'B', 'max_heart_rate': 200}  ... displaying 10 of 16 total bound parameter sets ...  {'athlete_id': 15, 'source_id': 'p15', 'age': 54, 'height_cm': 180, 'gender': 'male', 'personality_type': 'A', 'max_heart_rate': None}, {'athlete_id': 16, 'source_id': 'p16', 'age': 23, 'height_cm': 182, 'gender': 'male', 'personality_type': 'B', 'max_heart_rate': 199}]]
(Background on this error at: https://sqlalche.me/e/20/gkpj)